### PR TITLE
Improve serum liquidation

### PR DESCRIPTION
**Pull request by suscd**
Rework the calculations in the serum liquidation to find the worst acceptable price to set when placing a limit order. This allows for better efficiency by accepting trades with partial order fills.

This also now allows for serum liquidations to occur when the quote token is being used as collateral.

### CODE REVIEW

File: programs/jet/src/instructions/liquidate_dex.rs [hunk 1]
```diff
@@ -77,19 +77,34 @@ struct DexClient<'a, 'info> {
...
+                .as_u64_rounded(0),
+        )
+    }
+
     /// Buy as much of the base currency as possible with the given amount
     /// of quote tokens.
-    fn _buy(&self, quote_amount: u64) -> ProgramResult {
-        let limit_price = u64::MAX;
+    fn buy(&self, limit_price: u64, quote_amount: u64) -> ProgramResult {
         let max_coin_qty = u64::MAX;
```
[the remark this answers is not in the document]
Comment (suscd) [reply programs/jet/src/instructions/liquidate_dex.rs:100]: As discussed, adjusted the calculation to also set an upper bound based on the total debt for the position being liquidated.

File: programs/jet/src/instructions/liquidate_dex.rs [hunk 2]
```diff
@@ -77,19 +77,34 @@ struct DexClient<'a, 'info> {
...
         let max_pc_qty = quote_amount;
 
         self.create_order(Side::Bid, limit_price, max_coin_qty, max_pc_qty)
     }
 
     /// Sell as much of the given base currency as possible.
-    fn sell(&self, base_amount: u64) -> ProgramResult {
-        let limit_price = 1;
+    fn sell(&self, limit_price: u64, base_amount: u64) -> ProgramResult {
         let max_pc_qty = u64::MAX;
```
[the remark this answers is not in the document]
Comment (suscd) [reply programs/jet/src/instructions/liquidate_dex.rs:108]: As discussed, we can't bound both, and while in theory we could access the order book it's not really worth the effort right now. So for now we just accept there will be a bit of an approximation here, and we might buy more debt than needed (where any extra tokens are retained by the protocol)

File: programs/jet/src/instructions/liquidate_dex.rs [hunk 3]
```diff
@@ -297,14 +312,98 @@ impl<'info> LiquidateDex<'info> {
...
+        } else if c_ratio_ltv > min_c_ratio {
+            // This means the loan is underwater, so for now we just disallow
+            // liquidations on underwater loans using the DEX.
+            return Err(ErrorCode::Disallowed.into());
+        }
+
+        let limit_fraction = (c_ratio_ltv - Number::ONE)
+            / (min_c_ratio * (Number::ONE - liquidation_fee) - Number::ONE);
+
+        let collateral_sellable_value = limit_fraction * collateral_value;
```
[the remark this answers is not in the document]
Comment (suscd) [reply programs/jet/src/instructions/liquidate_dex.rs:470]: declined to change, as the existing calculation is proven to be accurate in practice

[the remark this answers is not in the document]
Comment (eyevz) [reply programs/jet/src/instructions/liquidate_dex.rs:470]: Okay I got to the bottom of this. Both the implemented expression and my suggested change are wrong. What's implemented systematically slightly over-estimates the amount of collateral that needs to be sold. It's okay not to change this now - we'll fix it when we refactor for margin accounts.

For completeness, the correct expression is

`limit_fraction = (1 - ltv * c_min) / (1 - c_min / (1 + liquidation_fee))`

**Review by eyevz — APPROVED** [review]

**Pull request merged**